### PR TITLE
bpo-43688: Support the limited C API in debug mode

### PR DESCRIPTION
**Pull request by vstinner**
The limited C API is now supported if Python is built in debug mode
(if the Py_DEBUG macro is defined). In the limited C API, the
Py_INCREF() and Py_DECREF() functions are now implemented as opaque
function calls, rather than accessing directly the PyObject.ob_refcnt
member, if Python is built in debug mode and the Py_LIMITED_API macro
targets Python 3.10 or newer. It became possible to support the
limited C API in debug mode because the PyObject structure is the
same in release and debug mode since Python 3.8 (see [bpo-36465](https://bugs.python.org/issue36465)).

The limited C API is still not supported in the --with-trace-refs
special build (Py_TRACE_REFS macro).

https://bugs.python.org/issue43688

### CODE REVIEW

File: Include/object.h [hunk 1]
```diff
@@ -427,21 +427,46 @@ PyAPI_FUNC(void) _Py_NegativeRefcount(const char *filename, int lineno,
...
+PyAPI_FUNC(void) Py_DecRef(PyObject *);
+
+// Similar to Py_IncRef() and Py_DecRef() but the argument must be non-NULL.
+// Private functions used by Py_INCREF() and Py_DECREF().
+PyAPI_FUNC(void) _Py_IncRef(PyObject *);
+PyAPI_FUNC(void) _Py_DecRef(PyObject *);
+
 static inline void _Py_INCREF(PyObject *op)
 {
+#if defined(Py_LIMITED_API) && (Py_LIMITED_API+0 >= 0x030A0000 || defined(Py_REF_DEBUG))
```
Comment (vstinner) [inline Include/object.h]: I would prefer to make Py_INCREF() opaque for Python 3.9 and older, but we cannot magically add _Py_IncRef() to old Python versions :-(

**Comment by vstinner** [comment]
cc @encukou

**Comment by vstinner** [comment]
I merged the uncontroversial part in separated commits in https://bugs.python.org/issue43688 I rebased my PR on top on merged changes.

File: setup.py [hunk 2]
```diff
@@ -1864,17 +1864,11 @@ def detect_modules(self):
...
-        else:
-            # Debug mode: Build xxlimited with the full API
-            # (which is compatible with the limited one)
-            self.add(Extension('xxlimited', ['xxlimited.c']))
-            self.add(Extension('xxlimited_35', ['xxlimited_35.c']))
+        # Limited C API
+        self.add(Extension('xxlimited', ['xxlimited.c'],
+                           define_macros=[('Py_LIMITED_API', '0x030a0000')]))
+        self.add(Extension('xxlimited_35', ['xxlimited_35.c'],
+                           define_macros=[('Py_LIMITED_API', '0x03050000')]))
```
Comment (shihai1991) [inline setup.py:1871]: Hm, I found `Py_LIMITED_API=0x03060000` in https://github.com/python/cpython/blob/master/PCbuild/xxlimited_35.vcxproj#L97.

I am not sure which one is right.

Comment (vstinner) [reply setup.py:1871]: The module name says "35": Python 3.5. See the C file header:

```

/* This module is compiled using limited API from Python 3.5,
 * making sure that it works as expected.
 *
 * See the xxlimited module for an extension module template.
 */
```

Comment (vstinner) [reply setup.py:1871]: Well, that's a good reason to put the define into the C file directly, rather than putting it in the recipe to build the file :-D

**Comment by vstinner** [comment]
Windows x86: test_asyncio failed.

**Comment by vstinner** [comment]
I re-run the CI jobs.

**Pull request merged**